mlflow/server/js/src/experiment-tracking/components/ExperimentViewUtil.js

```diff
@@ -637,23 +638,39 @@ export default class ExperimentViewUtil {
           expanderOpen: ExperimentViewUtil.isExpanderOpen(runsExpanded, runId),
           childrenIds,
           runId,
+          level: level
         },
       ];
     });
+    console.log("parent rows")
+    console.log(parentRows)
     const mergedRows = [];
+    const visited = new Set();
     parentRows.forEach((r) => {
-      const { runId } = r;
-      mergedRows.push(r);
-      const childrenIdxs = parentIdToChildren[runId];
-      if (childrenIdxs) {
-        if (ExperimentViewUtil.isExpanderOpen(runsExpanded, runId)) {
-          const childrenRows = childrenIdxs.map((idx) => {
-            return { idx, isParent: false, hasExpander: false };
-          });
-          mergedRows.push(...childrenRows);
+      if(r.level == 0 ){
+        function dfs(idx, curr_level){
+          if(!visited.has(idx)){
+            let row = parentRows.find(o => o.idx == idx)
+            if(!row){
+              row = { idx, isParent: false, hasExpander: false, level: curr_level};
+            }
+            mergedRows.push(row)
+            visited.add(idx)
+            const childrenIdxs = parentIdToChildren[row.runId];
+            if(childrenIdxs){
+              if (ExperimentViewUtil.isExpanderOpen(runsExpanded, row.runId)) {
+                childrenIdxs.forEach(idx => {
+                  dfs(idx, curr_level+1)
+                });
+              }
+            }
+          }
```

Comment:
I am curious if we can make the logic a little simple:

above we compute all the parentRows. can we compute mergedRows directly?

Where we can check if its a parent, then we set the children and the level correctly. 
If its a left, the children would be empty and the `hasExpander: false` and level would be set correctly. 

Not sure we need to check `r.level == 0` in the initial loop for `parentRows`. We are anyway going through all levels in the recursion loop. 

Let me know if I am missing something

Reply:
Yes I also think we can. 
After reading your comment, i refactored the code and made it simpler, and computed ```mergedRows``` directly.

The ```r.level == 0 ``` was to make sure that we only do the recursion loop on the root (which goes to children, and grandchildren, etc). If i didn't do this the order it would be incorrect, since ```runInfos``` is ordered by ```Start Time```.